src/man/man3/FollowWall.3
.TH FollowWall 3 "29 May 2002" "HulkRobot" \" -*- nroff -*-
.ad l
.nh
.SH NAME
FollowWall \- 
.SH SYNOPSIS
.br
.PP
\fC#include <FollowWall.h>\fR
.PP
.SS Public Methods

.in +1c
.ti -1c
.RI "\fBFollowWall\fR (int priority, int samplingPeriod, int stackSize, \fBgyro\fR *aG, \fBusonic\fR *aU, \fBwheel\fR *aW)"
.br
.ti -1c
.RI "\fB~FollowWall\fR ()"
.br
.ti -1c
.RI "int \fBinit\fR (void)"
.br
.ti -1c
.RI "int \fBstep\fR (void)"
.br
.ti -1c
.RI "int \fBrun\fR (void)"
.br
.ti -1c
.RI "int \fBstop\fR (void)"
.br
.in -1c
.SH CONSTRUCTOR & DESTRUCTOR DOCUMENTATION
.PP 
.SS FollowWall::FollowWall (int _priority, int _samplingPeriod, int stackSize, \fBgyro\fR * aG, \fBusonic\fR * aU, \fBwheel\fR * aW)
.PP
Definition at line 9 of file FollowWall.cc.
.PP
.nf
00011 {
00012 
00013   Error( ErrorDebug, "FollowWall initializing");
00014   priority        = _priority;
00015   samplingPeriod= _samplingPeriod;
00016   delayTicks      = samplingPeriod/OSTickPeriod;
00017 
00018   taskMode        = TaskStopped;
00019   taskSem = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00020   tickerSem       = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00021   taskExitSem     = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00022   tickerExitSem   = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00023 
00024   aGyro = aG;
00025   aUSonic = aU;
00026   aWheel = aW;
00027 
00028   Assert( taskSem && taskExitSem && tickerSem && tickerExitSem );
00029 
00030   taskId  = taskSpawn("FollowWall", priority, VX_FP_TASK, 
00031                           stackSize, (FUNCPTR) taskModuleStub,
00032                           (int) this, 0, 0, 0, 0, 0, 0, 0, 0, 0);
00033   Assert(taskId!=ERROR);
00034 
00035   semTake(taskExitSem, WAIT_FOREVER);
00036   
00037   tickerTaskId    = taskSpawn("FollowWall", TickerTaskPriority, 0, 
00038                           TickerTaskStackSize,
00039                           (FUNCPTR) taskModuleTickerStub,
00040                           (int) this,
00041                           0, 0, 0, 0,
00042                           0, 0, 0, 0, 0);
00043   Assert(tickerTaskId!=ERROR);
00044   semTake(tickerExitSem, WAIT_FOREVER);
00045 
.fi
.SS FollowWall::~FollowWall ()
.PP
Definition at line 47 of file FollowWall.cc.
.PP
.nf
00048 {
00049   taskMode=TaskTerminate;
00050   semGive(tickerSem);
00051   semGive(taskSem);
00052   semTake(tickerExitSem, WAIT_FOREVER);
00053   semTake(taskExitSem, WAIT_FOREVER);
00054   semDelete(taskSem);
00055   semDelete(taskExitSem);
00056   semDelete(tickerSem);
00057   semDelete(tickerExitSem);
00058 
00059   Error( ErrorDebug, "FollowWall destroyed");
00060 
.fi
.SH MEMBER FUNCTION DOCUMENTATION
.PP 
.SS int FollowWall::init (void)
.PP
Definition at line 183 of file FollowWall.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00184 {
00185 
.fi
.SS int FollowWall::run (void)
.PP
Definition at line 194 of file FollowWall.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00195 {
00196   taskMode=TaskRunning;
00197   semGive(tickerSem);
00198   return 0;
00199 
.fi
.SS int FollowWall::step (void)
.PP
Definition at line 187 of file FollowWall.cc.
.PP
.nf
00188 {
00189   taskMode=TaskStep;
00190   semGive(taskSem);
00191   return 0;
00192 
.fi
.SS int FollowWall::stop (void)
.PP
Definition at line 201 of file FollowWall.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00202 {
00203   taskMode=TaskStopped;
00204   return 0;
00205 
.fi


.SH AUTHOR
.PP 
Generated automatically by Doxygen for HulkRobot from the source code.

src/bbc/FollowWall.cc
#include "dfl.h"
#include "FollowWall.h"

unsigned FollowWall::forward;
unsigned FollowWall::middle;
unsigned FollowWall::reward;
unsigned FollowWall::difference;

FollowWall::FollowWall(int _priority, int _samplingPeriod, 
		int stackSize, gyro *aG, usonic *aU, wheel *aW)
{

  Error( ErrorDebug, "FollowWall initializing");
  priority 	= _priority;
  samplingPeriod= _samplingPeriod;
  delayTicks	= samplingPeriod/OSTickPeriod;

  taskMode	= TaskStopped;
  taskSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  tickerSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  taskExitSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  tickerExitSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);

  aGyro = aG;
  aUSonic = aU;
  aWheel = aW;

  Assert( taskSem && taskExitSem && tickerSem && tickerExitSem );

  taskId 	= taskSpawn("FollowWall", priority, VX_FP_TASK, 
				stackSize, (FUNCPTR) taskModuleStub,
				(int) this, 0, 0, 0, 0, 0, 0, 0, 0, 0);
  Assert(taskId!=ERROR);

  semTake(taskExitSem, WAIT_FOREVER);
  
  tickerTaskId	= taskSpawn("FollowWall", TickerTaskPriority, 0, 
				TickerTaskStackSize,
				(FUNCPTR) taskModuleTickerStub,
				(int) this,
				0, 0, 0, 0,
				0, 0, 0, 0, 0);
  Assert(tickerTaskId!=ERROR);
  semTake(tickerExitSem, WAIT_FOREVER);
}

FollowWall::~FollowWall()
{
  taskMode=TaskTerminate;
  semGive(tickerSem);
  semGive(taskSem);
  semTake(tickerExitSem, WAIT_FOREVER);
  semTake(taskExitSem, WAIT_FOREVER);
  semDelete(taskSem);
  semDelete(taskExitSem);
  semDelete(tickerSem);
  semDelete(tickerExitSem);

  Error( ErrorDebug, "FollowWall destroyed");
}

//
// C stub functions to call C++ member functions from
// a newly spawned task
//

void FollowWall::taskModuleStub(FollowWall *This, int d2, int d3, 
					int d4, int d5, int d6, int d7, 
					int d8, int d9, int d10)
{
  Error( ErrorDebug, "FollowWall taskModuleStub ");
  This->task();
}

void FollowWall::taskModuleTickerStub(FollowWall *This, int d2, 
			      int d3, int d4, int d5, int d6, int d7, 
			      int d8, int d9, int d10)
{
  Error( ErrorDebug, "FollowWall taskModuleTickerStub");
  This->ticker();
}

void FollowWall::ticker(void)
{
  semGive(tickerExitSem);
  while(taskMode!=TaskTerminate) {

    if(taskMode==TaskStopped || taskMode==TaskPaused) 
       semTake(tickerSem, WAIT_FOREVER);

    while(taskMode==TaskRunning) {
      Error( ErrorNone, "ticker::semGive" );
      semGive(taskSem);
      taskDelay(delayTicks);
    }
  }
  semGive(tickerExitSem);
  // task exit - deleted automatically
}

void FollowWall::task(void)
{
  semGive(taskExitSem); 		// signal that it has been created
  while(taskMode!=TaskTerminate) {
    semTake(taskSem, WAIT_FOREVER);	// wait until ticker ticks
    if (taskMode!=TaskRunning && taskMode!=TaskStep) continue;
    Error( ErrorNone, "task tick" );

    // task code HERE
    forward = aUSonic->usData(13);
    middle =  aUSonic->usData(12);
    reward =  aUSonic->usData(11);
    difference = abs(forward - reward);
    //printf(" [F] %d  [M], %d  [R],%d\n",forward,middle, reward);

    //
    // check side 3 sensors , configuration  [13]-[12]-[11]
    //

    //
    // basic rule
    //

    if(forward < reward) {
      turnMode = TURN_RIGHT;
      Error( ErrorNone, "TURN_RIGHT" );
    } 
    else if(forward>reward) {
      turnMode = TURN_LEFT;
      Error( ErrorNone, "TURN_LEFT" );
    }

    //
    // additional rule
    //

/*
    if(middle < SideRange) {
      turnMode = TURN_RIGHT;
      Error( ErrorNone, "TURN_RIGHT for minimize side" );
    }
*/

    //
    // for help MonitorDoor and MonitorCorner
    //

    if((forward>middle)&&(forward-middle)>100) {
      turnMode = TURN_STOP_SLOW;
      Error( ErrorNone, "TURN_STOP for help MonitorDoor" );
    }


    switch(turnMode) {
      case(TURN_RIGHT):
	aWheel->setSpeed(0.1);
	aWheel->setTurnSpeed(0.02);
	break;
      case(TURN_LEFT):
	aWheel->setSpeed(0.1);
	aWheel->setTurnSpeed(-0.02);
	break;
      case(TURN_STOP):
	aWheel->setSpeed(0.1);
	aWheel->setTurnSpeed(0.0);
	break;
      case(TURN_STOP_SLOW):
	aWheel->setSpeed(0.1);
	aWheel->setTurnSpeed(0.0);
	break;
    }
    // task code till HERE

  }
    // for stop when finish this task
	aWheel->setSpeed(0.0);
	aWheel->setTurnSpeed(0.0);

  semGive(taskExitSem); 
  // task exit - deleted automatically
}

int FollowWall::init(void)
{
}

int FollowWall::step(void)
{
  taskMode=TaskStep;
  semGive(taskSem);
  return 0;
}

int FollowWall::run(void)
{
  taskMode=TaskRunning;
  semGive(tickerSem);
  return 0;
}

int FollowWall::stop(void)
{
  taskMode=TaskStopped;
  return 0;
}

